### ISSUE TITLE
STY: Windows and temporary direcotories

### ISSUE THREAD

**Issue by jklenzing** [pysat/pysat#974]
**Is your feature request related to a problem? Please describe.**
Windows has issues sometimes cleaning up temporary directories.  Multiple errors including `PermissionError` are thrown at the same time. This largely affects how unit tests are run in the CI environment. Python 3.10 fixes this by adding the  `ignore_cleanup_errors` kwarg.

**Describe the solution you'd like**
Update usage of `tempfile.TemporaryDirectory` to include `ignore_cleanup_errors=True`.  In the meantime, a `try` / bare `except` is used around cleanup statements where known problems occur to keep the code running until older python support can be dropped (estimated to be sometime around the release of Python 3.12).

**Describe alternatives you've considered**
🚫 :window:

**Additional context**
See discussion in #966

**Comment by aburrell** [comment]
Is it bad that I really just want the alternative 👿

**Comment by jklenzing** [comment]
Not sure about bad, but normal might be a good description.

**Comment by rstoneback** [comment]
![](https://sayingimages.com/wp-content/uploads/let-the-hate-flow-through-you-emperor-palpatine-meme.jpg)

**Comment by rstoneback** [comment]
> Is it bad that I really just want the alternative 👿

No. Having fun with the emperor there. This week Dan Welling started posting on twitter on how he is moving to windows and that WSL makes it possible. Unfortunate reality is if we don't support windows then our potential user base is considerably smaller. I don't like it but as long as we don't have to do anything unreasonable to keep windows going then I'm gonna advocate to keep micro$oft support in. Emotionally though, I'm right there with ya.

**Comment by jklenzing** [comment]
Double-checking the relevant pull.  The syntax to update whenever python 3.9 is dropped is flagged with a TODO statement.

**Comment by rstoneback** [comment]
I may have found the source for this. With the move to multiple data directories, I needed a way to store which directory a given file set was in for the Files class. I added this to stored file information. When pysat's files class loaded this info, it assigned inst.files.data_path with the stored path, even if that path was not in pysat.params['data_dirs']. Anyway, the end effect here was previous temporary directories were being silently assigned to an Instrument, inconsistent with current settings in pysat.params['data_dirs']. So pysat was itself unintentionally inconsistent. There is a potential nominal fix in #983. Stored directories only assigned if in params['data_dirs']. I also upped the feedback to the logger.

**Comment by rstoneback** [comment]
See line in `test_utils_io` with TODO #974 to directly test for remaining open resources after loading file using xarray.

**Comment by jklenzing** [comment]
> Double-checking the relevant pull. The syntax to update whenever python 3.9 is dropped is flagged with a TODO statement.

Assuming the new fixes work on windows, we can close this.  Otherwise the 3.9 drop date is Apr 5, 2024 according to https://numpy.org/neps/nep-0029-deprecation_policy.html

**Comment by rstoneback** [comment]
We have a bandaid on the situation, but based on my poking there is an issue on windows with file handles/resources being left open. It isn't our code at issue, but leaving this open could help us remember to check for improvements in the future.

**Comment by jklenzing** [comment]
> Assuming the new fixes work on windows, we can close this. Otherwise the 3.9 drop date is Apr 5, 2024 according to https://numpy.org/neps/nep-0029-deprecation_policy.html

Adding that since we have the operational version at 3.6 (outside our control), we should not use the NEP29 schedule to drop this issue.

**Comment by JonathonMSmith** [comment]
I think the Windows github actions workflow is not using WSL. So maybe there's a future where we can drop windows CI and just make WSL a windows requirement

**Comment by aburrell** [comment]
For reference, WSL is [Windows Subsystem for Linux](https://learn.microsoft.com/en-us/windows/wsl/installhttps://learn.microsoft.com/en-us/windows/wsl/install) and not [World Surf League](https://www.worldsurfleague.com/)

**Comment by JonathonMSmith** [comment]
> For reference, WSL is [Windows Subsystem for Linux](https://learn.microsoft.com/en-us/windows/wsl/installhttps://learn.microsoft.com/en-us/windows/wsl/install) and not [World Surf League](https://www.worldsurfleague.com/)
🤙

**Comment by jklenzing** [comment]
> For reference, WSL is [Windows Subsystem for Linux](https://learn.microsoft.com/en-us/windows/wsl/installhttps://learn.microsoft.com/en-us/windows/wsl/install) and not [World Surf League](https://www.worldsurfleague.com/)

🏄
